Add Agentless Datadog Trace Exporter

ericmustin opened the pull request:
This is probably at a point where it could benefit from review, cc @mx-psi , some things to note:

- There's still some uncertainty around a few trace's span translation choices, as well as how to grab the global config stuff like hostname and env and apply it to the trace payloads (as well as the order of precedence for things like `DD_SERVICE` as opposed to the service already set by the tracer). but for the most part the implementation is complete, as is the flushing of trace and stats payloads.

- the most of what's in `/apm`, including tests and the snapshots, i've just grabbed [from serverless](https://github.com/DataDog/datadog-serverless-functions/tree/master/aws/logs_monitoring/trace_forwarder/internal/apm). I tried to remove everything that wasn't being used by our implementation, but didn't update the serverless tests, so they'll probably fail. 

- The big things here are going to be:
  - having a better set of eyes around my very hacky Go code. Anything you see that looks unusual, or could be optimised, there's likely _not_ a good reason i've chosen not to, it's just because i don't know Go especially well.
  - The implementation itself around converting Internal Traces to Datadog Traces largely follows the approach the [Zipkin Exporter](https://github.com/open-telemetry/opentelemetry-collector/tree/master/exporter/zipkinexporter) took. There are a few things zipkin is doing that we don't support and vice versa but the methods and code paths are similar.
  - The conventions of the coercing the internal spans to datadog spans largely follow the [python exporter's translation implementation](https://github.com/open-telemetry/opentelemetry-python/blob/b2559409b2bf82e693f3e68ed890dd7fd1fa8eae/exporter/opentelemetry-exporter-datadog/src/opentelemetry/exporter/datadog/exporter.py#L229), while the conventions and approach around generating+aggregating+flushing trace+stats payloads is largely based on the [serverless implementation](https://github.com/DataDog/datadog-serverless-functions/tree/master/aws/logs_monitoring/trace_forwarder/internal/apm). As i mentioned previously, I more or less ported everything in `/apm` directly from that repo.
  - I've been testing this locally using our `otel-collector-e2e-tests` package, lmk if you have any trouble with that.
  - How to appropriately manage the config. it's possible my approach has drifted from what we've ended up doing in the metric exporter.
  - Tests are going to be a hassle, If you see any shortcuts here or good similar test setups in other parts of the repo that I can use as a guide, let me know. I will try to start writing tests for this stuff next

ericmustin committed:
[wip] add client to exporter

ericmustin committed:
truncate attrib strings

ericmustin committed:
[wip] add stats payload begin env and hostname work

ericmustin committed:
[wip]: add rough hostname resolution

ericmustin committed:
[wip] trace translation work

ericmustin committed:
add obfuscation

ericmustin committed:
[wip] migrate span translation to otel internals

ericmustin committed:
[wip] span translation

ericmustin committed:
[wip] use passed in hostname and env from config and also remove unused serverless code

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+
+// ddSpan represents the Datadog span definition.
+type ddSpan struct {
+	SpanID   uint64             `msg:"span_id"`
```
I don't think we need the `msg` struct tags here: they are used for MessagePack which we are not using (right?). We can add them back in the future if we add support for sending traces to the Trace Agent.

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+	// get env tag
+	env := cfg.Env; if env == "" {
+		env = "none"
+	}
```
Should we handle this during config sanitizing? I think it would make sense, since this should be handled for host tags too

mx-psi replied:
I see this was removed but not added to `Config.Sanitize`, was this intended?

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+	}
+
+	// group the traces by env to reduce the number of flushes
+	return aggregateTracePayloadsByEnv(traces), nil
```
I think we could simplify this function if we move this last bit outside of the function. That way we can use `traces` as named result parameter and avoid the `if resourceSpans.Len() == 0` bit.

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+	resource := rs.Resource()
+	ilss := rs.InstrumentationLibrarySpans()
+
+	// TODO: should we be getting env from the trace's spans itself? if so what's the tag+precedence order?
```
Is this generally set by instrumentation libraries? And I am assuming this is set as an attribute, right?

ericmustin replied:
I believe it would be an attribute, probably on the resource? but i dont think anything is standard, we should probably defer to `env` or `environment` attribute on either the resource, instrumentation library, or span level, but going to try to see if there's anything in spec on this

ericmustin replied:
so i spoke with one of the tracing client maintainers, there's a non-required resource convention `deployment.environment` https://github.com/open-telemetry/opentelemetry-specification/blob/4b6f6e502e48e49b3bca93b13f309f9a61d5e1f7/specification/resource/semantic_conventions/deployment_environment.md#deployment that seems like the 'official otel' way of doing env. so let's defer to that, but if it doesn't exist, use DD_ENV / our env config

mx-psi replied:
That sounds good to me, marking this as solved

mx-psi replied:
Since this is solved we should remove the comment
```suggestion
```

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+
+			if err != nil {
+				return payload, err
+			} else {
```
nit: This `else` is unnecessary since we return on the `if` branch

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+						TraceID:   span.TraceID,
+						Spans:     []*pb.Span{},
+						StartTime: 0,
+						EndTime:   0,
```
I found this confusing; maybe add a comment here saying that this time is handled on `addToAPITrace`?

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+	endTime := s.EndTime()
+	duration := int64(endTime) - int64(startTime)
+
+	// it's possiblle end time is unset, so default to 0 rather than using a negative number
```
```suggestion
	// it's possible end time is unset, so default to 0 rather than using a negative number
```

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+
+
+// convertSpan takes an internal span representation and returns a Datadog span.
+func spanToDatadogSpan(	s pdata.Span,
```
You should run `make fmt` to format the code

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+
+	// it's possiblle end time is unset, so default to 0 rather than using a negative number
+	if s.EndTime() == 0 {
+		duration = int64(0)
```
I think we can do
```suggestion
		duration = 0
```
can't we?

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+		SpanID:  decodeAPMId(hex.EncodeToString(s.SpanID().Bytes()[:])),
+		Name:     getDatadogSpanName(s, tags),
+		Resource: getDatadogResourceName(s, tags),
+		Service:  "service_example",
```
We should get the service from `TagsConfig.Service`.

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+		} else {
+			span.Error = 0
+		}
+
```
nit: we can simplify this since `span.Error` is set to 0 by default (or leave it /a comment for emphasis)
```suggestion
		}
```

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,517 @@
+		tags[httpStatusCode] = status.Code().String()
+	} else {
+		span.Error = 0
+	}
```
ditto
```suggestion
	}
```

mx-psi reviewed (commented):
I reviewed some parts of the code and left a few comments so that you can keep working on this; most importantly I think you need to run `make fmt` and `make` to run all the linters and tests that can help catch errors (we will need them to pass to have a PR accepted upstream)

ericmustin committed:
linting'

ericmustin committed:
impement feedback first round

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,506 @@
+		}
+
+		// TODO: Also pass in globalTags here when we know what to do with them
+		payload, err := resourceSpansToDatadogSpans(rs, hostname, env, service, version)
```
nit: It makes more sense to pass the whole `TagsConfig` struct here since we will be using most if not all fields.

ericmustin replied:
gotcha, makes sense will update

mx-psi commented on exporter/datadogexporter/traces_exporter.go:
```diff
@@ -16,25 +16,92 @@ package datadogexporter
+
+		// this is for generating metrics like hits, errors, and latency, it uses a seperate endpoint than Traces
+		stats := apm.ComputeAPMStats(ddTracePayload)
+		err_stats := exp.edgeConnection.SendStats(context.Background(), stats, 3)
```
Can we get away with using OpenTelemetry retries here? If `pushTraceData` returns an error the exporter helper will deal with retrying. If we could avoid the `SendStats` and `SendTraces` using retries we could simplify this code a lot. 

The problem here would be if e.g. `SendStats` fails but `SendTraces` does not. We would have to signal failure here and so we would be sending traces twice. Would the endpoint be okay with that (and with the reversed case)?

ericmustin replied:
i believe duping 1/2 would be a possibility here, the endpoint is pretty naive iirc. There's a note in the serverless repo on this. 

Thinking about it though, it probably makes sense to prioritize `SendTraces` imo, since metrics are based on the traces, so if there's something unaccepted with the traces we probably don't want to trust the metrics that are generated anyway

ericmustin replied:
so, looked through this again and i think we'd have issues since we're exporting potentially multiple `ddTracePayload` since we iterate over the `range aggregatedTraces`. It's one payload per env. So, tl;dr I don't think we could easily support using Otel default retries here, or at least i'm not quite how we'd handle it right now

mx-psi replied:
Okay, guess we will have to deal with it :( I still think we could benefit from having a function like

 ```go
func pushWithRetry(maxRetries int, waitTime time.Duration, fn func() error) error {
  // ...
}
```

that abstracts the retrying logic in case we want to change it afterwards, but feel free to mark this as done if you think that's not worth the effort.

ericmustin replied:
yea i think that's reasonable, let me see if i can make this a bit cleaner

ericmustin replied:
I tried to add at least the structure of how we'd use a retry  func here. wiring up all the wait Duration retry stuff seemed a bit complicated so i passed on that for the moment, but i think we now at least have the structure where we can just update the error handling function if we want to handle retries differently.

I also dropped the retry amount to 0 (passing in 1), since it's unclear to me if we dedupe in most cases, so would rather not experiment, if we experience lots of drops we can iterate.

mx-psi commented on exporter/datadogexporter/internal/apm/model.go:
```diff
@@ -0,0 +1,175 @@
+
+// AddTagsToTracePayloads takes a single string of tags, eg 'a:b,c:d', and applies them
+// to each span in a trace
+func AddTagsToTracePayloads(tracePayloads []*pb.TracePayload, tags string) {
```
We can make this private (lowercase) since it is only used in tests.

ericmustin replied:
I just went ahead and removed this along with the tests...i think we're gonna have to write these tests from scratch since we're dealing with a pretty significantly pared down version of the serverless code

mx-psi commented on exporter/datadogexporter/internal/apm/model.go:
```diff
@@ -0,0 +1,175 @@
+				}
+			}
+		}
+	}
```
We probably don't need this, right? Although removing it will probably break tests of the serverless part :(

ericmustin replied:
yup, removed

mx-psi commented on exporter/datadogexporter/internal/apm/trace-connection.go:
```diff
@@ -0,0 +1,202 @@
+/*
```
Name of the file should be in snake case, as in `trace_connection.go`

ericmustin replied:
👍 updated

mx-psi commented on exporter/datadogexporter/internal/apm/model.go:
```diff
@@ -0,0 +1,175 @@
+/*
```
I am not sure if upstream reviewers would be okay with having our copyright headers here :/

ericmustin replied:
whoops, updated

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,506 @@
+				message: "ERR_CODE_" + strconv.FormatInt(int64(status.Code()), 10),
+				status:  http.StatusInternalServerError,
+			}
+		}
```
Why do we assume an internal server error if the status code is not found in our map?

ericmustin replied:
🤷  just borrowing from the open census code (i think?) here. let me see if there's a more reasonable default

ericmustin replied:
so looking into the collector a bit this seems to be what folks are using for "unknown error", so i think it's a reasonable default imo...we can see what they say in review.

mx-psi replied:
Okay, if that's what others are doing it makes sense to me!

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,506 @@
+
+	// Set Attributes as Tags
+	for key, val := range tags {
+		setTag(span, key, val)
```
Since we know the `val`s are strings, wouldn't it make more sense to use `setStringTag` here? And then I am  not sure what do we have `setTag` for, is it for the commented code above?

ericmustin replied:
this is sorta a frankenstein left over from converting from open census...i'll try to clean this up

ericmustin replied:
updated this to just rely on setStringTag

mx-psi reviewed (commented):
Second round; `impi` (after running `make fmt`) and `misspell` still give some errors that need to be addressed

ericmustin committed:
linting, package cleanup

ericmustin committed:
clean up tag mappings and pass in whole config instead of specific tags

NathanBaulch commented:
Well that's annoying, I just stumbled on this branch/PR after successfully migrating the tracing portion of `opencensus-go-exporter-datadog` to OpenTelemetry [here](https://github.com/NathanBaulch/opencensus-go-exporter-datadog/commit/4024a9d69c078a1d567c548fcc540dda4f69e256). I admit it did seem weird that this work hadn't been done already (at least not somewhere easily discoverable).

Oh well, mine works but is just a minimal 1-1 (where possible) port, so I look forward to this official one being released. :+1:

While I'm here (and sorry if off-topic), I'm curious about the pros and cons of an OTEL exporter vs an OTEL API compatible wrapper as used in `dd-trace-go` ([example](https://github.com/DataDog/dd-trace-go/tree/v1/ddtrace/opentracer))?

Also, does "agentless" mean it'll go direct to datadoghq without dd-agent? Thanks.

mx-psi commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,208 @@
+	// communicating with Datadog
+	var userAgent = fmt.Sprintf(
+		"%s/%s/%s (+%s)",
+		"xray-converter", "0.1", "1", "http://localhost",
```
Does this user agent make sense for Otel too? `xray-converter` seems to refer to AWS X-Ray

ericmustin replied:
ah good catch, i'll update (will try to check if somethiing on the backend relies on this)

mx-psi commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,208 @@
+		}
+
+		// All others aren't
+		return err
```
I know this code isn't yours but... if the error isn't retriable, shouldn't we return `nil` here? 

Thinking a bit about this I don't think it's necessarily the case that 4XX errors are not retriable e.g. if you are rate limited so I am even more confused by this

mx-psi commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,208 @@
+
+	if err != nil {
+		return err
+	}
```
Not sure if we want to retry if we `err` here, if we can't build the request object it means something is wrong in our code (like, we used an invalid method or we had a malformed URL), I think we should log the error and return `nil`

ericmustin commented:
@NathanBaulch the goal here is we really need to be able to live in the otel-collector. simply put, existing otel users likely aren't going to want to run a datadog agent along with an opentelemetry-collector (maybe they will? we plan to add agent export as well down the road). So, yes agentless means direct to api intake. We're still iterating here, as circumventing the agent and our own tracing clients introduces a whole can of worms around metric computation but we hope to have a PR in with the Otel folks by EOW. 

as far as `the pros and cons of an OTEL exporter vs an OTEL API compatible wrapper`, i think it's different use cases, an OTEL Exporter that lives in the opentelemetry-collector would be language agnostic but require the opentelemetry collector to be running, whereas the OTEL API Compatibility wrapper would leverage existing DD tracing architecture.

Sorry to hear about the duped work (nice work tho, fwiw!), but we can certainly keep you updated here on how this work is going and we'd love any feedback.

mx-psi commented on exporter/datadogexporter/stats.go:
```diff
@@ -0,0 +1,72 @@
+				Weight:   1,
+				TopLevel: true,
+			}
+			statsRawBucket.HandleSpan(weightedSpan, tracePayload.Env, []string{}, sublayers)
```
We should double check if this applies to Otel and update the comment

mx-psi commented on exporter/datadogexporter/traces_exporter.go:
```diff
@@ -16,25 +16,91 @@ package datadogexporter
+	aggregatedTraces := aggregateTracePayloadsByEnv(ddTraces)
+
+	// security/obfuscation for db, query strings, stack traces, pii, etc
+	// TODO: is there any config we want here? OTEL has their own pipeline for regex obfuscation
```
What pipeline does OTEL have?

ericmustin replied:
There's some span processors that allow regex to modify/filter specific attribs. 

- https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor/attributesprocessor
- https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor/spanprocessor

It's independent from a datadog exporters but i think it would be nice to eventually have documentation of suggested best practices for pipeline setups, or even plug n' play yaml examples for different use cases that we could have live in either a datadog exporter specific readme or docs section.

mx-psi replied:
Okay! I created OTEL-19 internally to keep track of this; we can mark as solved

mx-psi commented on exporter/datadogexporter/model.go:
```diff
@@ -0,0 +1,116 @@
+
+import (
+	// "strings"
+
```
```suggestion
```

mx-psi commented on exporter/datadogexporter/model.go:
```diff
@@ -0,0 +1,116 @@
+	return top
+}
+
+func ComputeSublayerMetrics(t pb.Trace) {
```
I am not sure what this function does, should we add a comment? Maybe it's obvious coming from an APM background but from an outsider's point of view it's unclear

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,459 @@
+		span.ParentID = idVal
+	} else {
+		span.ParentID = 0
+	}
```
You can do
```suggestion
	if len(s.ParentSpanID().Bytes()) > 0 {
		span.ParentID = decodeAPMId(hex.EncodeToString(s.ParentSpanID().Bytes()[:]))
	} 
```

ericmustin replied:
makes sense, updated

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -0,0 +1,459 @@
+	datadogType := spanKindToDatadogType(s.Kind())
+
+	span := &pb.Span{
+		TraceID:  decodeAPMId(hex.EncodeToString(s.TraceID().Bytes()[:])),
```
nit: We always use `decodeAPMId` after calling `hex.EncodeToString(id.Bytes()[:])` for a certain `id`, maybe we should move that inside `decodeAPMId` too

ericmustin replied:
updated this to move the `hex.encodeToString` bit within the helper func. i prefer to leave it as accepting a byte array rather than a specific span data field, i think it makes testing a bit easier.

mx-psi reviewed (commented):
Looking better! I left some more comments, I think we would benefit from some unit test coverage at this point. One thing I am worried about is that the Agent might spam logs when importing its packages/running its internal functions, have you noticed anything in the end to end tests?

ericmustin committed:
improve retry logic

ericmustin committed:
start specing out tests

ericmustin committed:
add more robust basic trace translation test

ericmustin committed:
add test for resource and error handling'
'

ericmustin committed:
add tests for translation helper funcs

ericmustin committed:
add retry helper functions, reduce retry, remove unused code

ericmustin committed:
remove unused mocks for trace exporting

ericmustin commented:
@mx-psi I tried to address most of the feedback and add unit tests for the span/trace translation work. Some of the trace connection stuff is still not covered in tests, I think I have to wire up a mock server for some of that, and I wasn't totally sure how to manage merging up the config tags. But punting on those bits for now so I can work on rebasing off the most up to date metrics exporter work so that the config is in a good place. I'd also like to start gameplanning on how to start the upstream PR process if you think this is in a reasonably good place, so let me know if there are other areas here where the code needs to be fixed before that.

ericmustin committed:
brig up to date with master

ericmustin committed:
fix tests and handle global tags

ericmustin committed:
add basic trace exporter tests

ericmustin commented:
Added some additional tests, rebasing off master brought in a bunch of changes to metric stuff that probably is now out of date. Going to wait for your metrrics PR and try to rebase off that branch

ericmustin committed:
merge upstream master

ericmustin committed:
run make

ericmustin committed:
try to bring up to date with traces/dev

ericmustin committed:
clean up factory test

ericmustin committed:
newline
'

ericmustin committed:
linting

ericmustin committed:
add factory test'

Commit:
Use fixed, mock credentials in emfexporter test. (#1176)

Set environment variables for AWS credentials to mock values. From what I can tell, the test fails if someone has valid AWS credentials on their machine currently, so to make behavior consistent we need the test to setup fixed creds.

Fixes #1149

asuresh4 committed:
Align Memory limit for TestTrace10kSPS with core (#1183)

Matches limit in core: https://github.com/open-telemetry/opentelemetry-collector/blob/master/testbed/tests/trace_test.go#L64

asuresh4 committed:
Update core to latest (#1159)

jpkrohling committed:
Fixed typo in systemd service file (#1180)

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

nebffa committed:
Allow MSI upgrades. (#1165)

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1164

tigrannajaryan committed:
Update core to latest version (#1172)

Had to fix SpanID usage since it is changed in the core.

ericmustin committed:
add Datadog Trace Export Configuration (#1181)

* add datadog tracer export config

* linting datadogexporter

ericmustin committed:
add trace connecction test with mock server

ericmustin committed:
fix span naming wonkiness and more robustly test trace connection

ericmustin commented:
@mx-psi So i think this is pretty much in a good state, i added mock server tests to test the trace connection so i think our code coverage is pretty good at this point. I just need to rebase off master + your metric exporter work

mx-psi commented on exporter/datadogexporter/model.go:
```diff
@@ -0,0 +1,91 @@
+
+	// iterate on each span and mark them as top-level if relevant
+	for _, span := range sps {
+		// The tracer can mark a span to be analyzfed, with a value 0-1, where 1 is always keep, and 0 is always reject.
```
```suggestion
		// The tracer can mark a span to be analyzed, with a value 0-1, where 1 is always keep, and 0 is always reject.
```

mx-psi commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -93,14 +99,18 @@ func (con *traceEdgeConnection) SendTraces(ctx context.Context, trace *pb.TraceP
+
+		if shouldRetry {
+			time.Sleep(traceEdgeRetryInterval)
+		}
```
Not that it matters if we are only retrying once but (morally) shouldn't we break out of the loop if `!shouldRetry`?

ericmustin replied:
good catch, yup will update

mx-psi commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -128,32 +138,35 @@ func (con *traceEdgeConnection) SendStats(ctx context.Context, sts *stats.Payloa
-		time.Sleep(traceEdgeRetryInterval)
+		if shouldRetry {
+			time.Sleep(traceEdgeRetryInterval)
+		}
```
ditto

mx-psi commented on exporter/datadogexporter/traces_exporter.go:
```diff
@@ -18,46 +18,100 @@ import (
+	// Calculate tags at startup
+	tags := cfg.TagsConfig.GetTags(false)
+	// TODO:
+	// use passed in config values for site and api key instead of hardcoded
```
I think this is fixed
```suggestion
```

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -82,159 +69,298 @@ var statusCodes = map[int32]codeDetails{
+
+	for _, apiTrace := range apiTraces {
+		// calculates analyzed spans for use in trace search and app analytics
+		// appends a specific piecce of metadata to these spans marking them as analyzed
```
```suggestion
		// appends a specific piece of metadata to these spans marking them as analyzed
```

mx-psi commented on exporter/datadogexporter/translate_traces.go:
```diff
@@ -82,159 +69,298 @@ var statusCodes = map[int32]codeDetails{
+}
+
+// TODO: determine why this always seems to be SPAN_KIND_UNSPECIFIED
+// even though span.kind is getting set at the app tracer level
```
Did you find out why this was happening? A bug on the receiver/client end?

ericmustin replied:
i believe it's related to the jThrift span translation code with the jaeger receiver used in our e2e tests, but i haven't have a chance to create a reproduction (its possibly they python client too). I'm updating this note for future reference

mx-psi reviewed (approved):
Looks good, I have yet to test it on the backend but since you already did I am going to approve. I left a bunch of small typos to fix/TODOs that can be removed too.

ericmustin committed:
merge master

ericmustin committed:
merge master

ericmustin committed:
clear up differences vs master

ericmustin committed:
add better code comments

dependabot[bot] committed:
Bump github.com/klauspost/compress in /exporter/honeycombexporter (#1185)

Bumps [github.com/klauspost/compress](https://github.com/klauspost/compress) from 1.11.0 to 1.11.1.
- [Release notes](https://github.com/klauspost/compress/releases)
- [Changelog](https://github.com/klauspost/compress/blob/master/.goreleaser.yml)
- [Commits](https://github.com/klauspost/compress/compare/v1.11.0...v1.11.1)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/aws/aws-sdk-go in /receiver/awsxrayreceiver (#1186)

Bumps [github.com/aws/aws-sdk-go](https://github.com/aws/aws-sdk-go) from 1.34.33 to 1.35.2.
- [Release notes](https://github.com/aws/aws-sdk-go/releases)
- [Changelog](https://github.com/aws/aws-sdk-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-go/compare/v1.34.33...v1.35.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/aws/aws-sdk-go in /internal/awsxray (#1189)

Bumps [github.com/aws/aws-sdk-go](https://github.com/aws/aws-sdk-go) from 1.34.32 to 1.35.2.
- [Release notes](https://github.com/aws/aws-sdk-go/releases)
- [Changelog](https://github.com/aws/aws-sdk-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-go/compare/v1.34.32...v1.35.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/aws/aws-sdk-go in /internal/awsxray/testdata/sampleapp (#1193)

Bumps [github.com/aws/aws-sdk-go](https://github.com/aws/aws-sdk-go) from 1.34.33 to 1.35.2.
- [Release notes](https://github.com/aws/aws-sdk-go/releases)
- [Changelog](https://github.com/aws/aws-sdk-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-go/compare/v1.34.33...v1.35.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/signalfx/sapm-proto in /receiver/sapmreceiver (#1195)

Bumps [github.com/signalfx/sapm-proto](https://github.com/signalfx/sapm-proto) from 0.6.0 to 0.6.2.
- [Release notes](https://github.com/signalfx/sapm-proto/releases)
- [Commits](https://github.com/signalfx/sapm-proto/compare/v0.6.0...v0.6.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/shirou/gopsutil in /exporter/signalfxexporter (#1187)

Bumps [github.com/shirou/gopsutil](https://github.com/shirou/gopsutil) from 2.20.8+incompatible to 2.20.9+incompatible.
- [Release notes](https://github.com/shirou/gopsutil/releases)
- [Commits](https://github.com/shirou/gopsutil/compare/v2.20.8...v2.20.9)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/shirou/gopsutil in /extension/observer/hostobserver (#1198)

Bumps [github.com/shirou/gopsutil](https://github.com/shirou/gopsutil) from 2.20.6+incompatible to 2.20.9+incompatible.
- [Release notes](https://github.com/shirou/gopsutil/releases)
- [Commits](https://github.com/shirou/gopsutil/compare/v2.20.6...v2.20.9)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/aws/aws-sdk-go in /processor/resourcedetectionprocessor (#1196)

Bumps [github.com/aws/aws-sdk-go](https://github.com/aws/aws-sdk-go) from 1.34.32 to 1.35.2.
- [Release notes](https://github.com/aws/aws-sdk-go/releases)
- [Changelog](https://github.com/aws/aws-sdk-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-go/compare/v1.34.32...v1.35.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ericmustin committed:
add back env var work

mx-psi committed:
[datadogexporter] Implement translation of metrics to Datadog format (#1178)

* Add metrics translation functions

* Improve docs and make newGauge private

* Add basic hostname resolution
This will be expanded in the future to more closely replicate the
Datadog Agent behavior

* Remove Series struct and use []datadog.Metric
We can add it back if necessary in the future

ericmustin committed:
merge upstream master

ericmustin committed:
merge datadog metric work updates

ericmustin committed:
remove unrelated metrics work

ericmustin committed:
[datadogexporter] add more unit tests for translation

ericmustin committed:
testing out stat bucket duration

ericmustin committed:
update to use an assumed 10s stats bucket

ericmustin committed:
update readme to include batch example

ericmustin committed:
remove batch from testdata

mx-psi commented:
Should we close this now that it is available upstream?
